lib/SIL/Utils/BasicBlockUtils.cpp

```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
       ReachableBlocks.insert(Pred);
   }
 }
+
+//===----------------------------------------------------------------------===//
+//                    Post Dominance Computation Utilities
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct PostDominanceCompletionSetComputer {
+  SILBasicBlock *definingBlock;
+  SILBasicBlock *reachableBlock;
+  SmallPtrSetImpl<SILBasicBlock *> &visitedBlocks;
+  DeadEndBlocks &deBlocks;
+
+  SmallVector<SILBasicBlock *, 32> worklist;
+  SmallSetVector<SILBasicBlock *, 8> successorBlocksThatMustBeVisited;
+
+  /// Set to true if we found a loop while processing.
+  bool foundLoop = false;
+
+  PostDominanceCompletionSetComputer(
+      SILBasicBlock *definingBlock, SILBasicBlock *reachableBlock,
+      SmallPtrSetImpl<SILBasicBlock *> &visitedBlocks,
+      DeadEndBlocks &deadEndBlocks)
+      : definingBlock(definingBlock), reachableBlock(reachableBlock),
+        visitedBlocks(visitedBlocks), deBlocks(deadEndBlocks) {
+    assert(definingBlock != reachableBlock);
+  }
+
+  void performDataflow();
+  void initializeDataflow();
+};
+
+} // anonymous namespace
+
+void PostDominanceCompletionSetComputer::initializeDataflow() {
```

Comment:
This algorithm does not have a separate "initialization" step, which implies something separate from the worklist processing. This function should just be a `pushPredecessors` utility called by the main `performDataflow`, once before entering the loop and once inside. It's actually alarming that it doesn't do the same thing, e.g. mark the predecessor visited.